text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+
+  /**
+   * The signing protocol
+   */
+  public readonly protocol: SigningProtocol;
+  /**
+   * Which requests CloudFront signs.
+   */
+  public readonly behavior: SigningBehavior;
+
+  public constructor(protocol: SigningProtocol, behavior: SigningBehavior) {
+    this.protocol = protocol;
+    this.behavior = behavior;
+  }
+}
+
+/**
+ * An Origin Access Control.
+ * @resource AWS::CloudFront::OriginAccessControl
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudfront-originaccesscontrol.html
+ */
+export class OriginAccessControl extends Resource implements IOriginAccessControl {
+  /**
+   * Imports an origin access control from its id and origin type.
+   */
+  public static fromOriginAccessControlAttributes(scope: Construct, id: string, attrs: OriginAccessControlAttributes): IOriginAccessControl {
+    class Import extends Resource implements IOriginAccessControl {
+      public readonly originAccessControlId = attrs.originAccessControlId;
+      public readonly originAccessControlOriginType = attrs.originAccessControlOriginType;
+    }
+    return new Import(scope, id);
+  }
+  /**
+   * Creates an origin access control for a S3 origin.
+   */
+  public static forS3(scope: Construct, id: string, props?: CommonOriginAccessControlProps): IOriginAccessControl {
+    return new OriginAccessControl(scope, id, {
+      ...props,
+      originAccessControlOriginType: OriginAccessControlOriginType.S3,
+    });
+  }
+
+  public readonly originAccessControlId: string;
+  public readonly originAccessControlOriginType: string;
+
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps) {
+    super(scope, id);
+    this.originAccessControlOriginType = props.originAccessControlOriginType;
+
+    // Check if origin access control name is 64 characters or less
+    if (props.originAccessControlName) {
+      this.validateOriginAccessControlName(props.originAccessControlName);
+    }
+
+    const resource = new CfnOriginAccessControl(this, 'Resource', {
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? Names.uniqueResourceName(this, {
+          maxLength: 64,
+        }),
+        signingBehavior: props.signing?.behavior ?? SigningBehavior.ALWAYS,
+        signingProtocol: props.signing?.protocol ?? SigningProtocol.SIGV4,
+        originAccessControlOriginType: this.originAccessControlOriginType,
+      },
+    });
+
+    this.originAccessControlId = resource.attrId;
+  }
+
+  private validateOriginAccessControlName(name: string) {
+    if (!Token.isUnresolved(name) && name.length > 64) {
+      throw new Error(`Origin access control name must be 64 characters or less, '${name}' has length ${name.length}`);
+    }
+  }
+}
+```
+
+#### Modifications to `S3BucketOrigin` class
+
+The `S3BucketOrigin` will have two methods, `withAccessIdentity()` and `withAccessControl()`, which each return a class configured with
```

Comment:
Please show an example of their usage immediately after their introduction, it's hard to validate this as is. I don't believe that this is the right approach. Consider the developer experience of this:

1. A user wants a `accessControl`, so they create a `BucketOrigin`. 
2. When they look at the code completion and the documentation, they become very confused because some properties apply to access identity and others apply to access control.
3. They accidentally specify some conflicting props, and then have to waste time correcting them. 
4. Now that they've configured this thing, they still have to manually specify `withAccessControl()` whereever it's used. If they type `withAccessIdentity()` by mistake, they will at best get an error, and at worst get undefined / unexpected behavior. 

You're telling the user to configure this one monolith class, but whenever they use it, to consciously choose the subclass to pass. 

This is not the way to do it. Create the two separate classes and deprecate the original monolith, or do the method that Kendra